lib/membrane_mp4/demuxer/isom/sample_data.ex

```diff
@@ -0,0 +1,193 @@
+defmodule Membrane.MP4.Demuxer.ISOM.SampleData do
+  @moduledoc false
+  alias Membrane.{Buffer, Time}
+  alias Membrane.MP4.Container
+  alias Membrane.MP4.MovieBox.SampleTableBox
+  alias Membrane.MP4.Track.SampleTable
+
+  @enforce_keys [
+    :samples,
+    :tracks_number,
+    :timescales,
+    :last_dts,
+    :sample_tables
+  ]
+
+  defstruct @enforce_keys
+
+  @type t :: %__MODULE__{
+          samples: [
+            %{
+              size: pos_integer(),
+              sample_delta: pos_integer(),
+              track_id: pos_integer()
+            }
+          ],
+          timescales: %{
+            (track_id :: pos_integer()) => timescale :: pos_integer()
+          },
+          last_dts: %{
+            (track_id :: pos_integer()) => last_dts :: Ratio.t() | nil
+          },
+          tracks_number: pos_integer(),
+          sample_tables: %{(track_id :: pos_integer()) => SampleTable.t()}
+        }
+
+  @spec get_samples(t, data :: binary()) ::
+          {[{Buffer.t(), track_id :: pos_integer()}], rest :: binary, t}
+  def get_samples(sample_data, data) do
+    {sample_data, rest, buffers} = do_get_samples(sample_data, data, [])
+
+    {buffers, rest, sample_data}
+  end
+
+  defp do_get_samples(%{samples: []} = sample_data, data, buffers) do
+    {sample_data, data, Enum.reverse(buffers)}
+  end
+
+  defp do_get_samples(sample_data, data, buffers) do
+    [%{size: size, track_id: track_id} = sample | samples] = sample_data.samples
+
+    if size <= byte_size(data) do
+      <<payload::binary-size(size), rest::binary>> = data
+
+      {dts, sample_data} = get_dts(sample_data, sample)
+
+      buffer =
+        {%Buffer{
+           payload: payload,
+           dts: dts
+         }, track_id}
+
+      sample_data = %{sample_data | samples: samples}
+      do_get_samples(sample_data, rest, [buffer | buffers])
+    else
+      {sample_data, data, Enum.reverse(buffers)}
+    end
+  end
+
+  defp get_dts(sample_data, %{sample_delta: delta, track_id: track_id}) do
+    use Ratio
+
+    dts =
+      case sample_data.last_dts[track_id] do
+        nil ->
+          0
+
+        last_dts ->
+          last_dts +
+            delta / sample_data.timescales[track_id] *
```

Comment:
shouldn't we add a delta of the previous sample, not the current sample?

Reply:
The delta of the sample describes when the sample should be decoded after the previous one. So this is ok.